std/selector/multiplexer.go

```diff
@@ -105,6 +98,39 @@ func generateSelector(api frontend.API, wantMux bool, sel frontend.Variable,
 	return out
 }
 
+// BinaryMux is an n to 1 multiplexer which uses a binary selector. selBits are
+// the selector bits, and the input at the index equal to the binary number
+// represented by the selector bits will be selected. More precisely the output
+// will be:
+//
+//	inputs[selBits[0]+selBits[1]*(1<<1)+selBits[2]*(1<<2)+...]
+//
+// len(inputs) does not need to be necessarily a power of 2. However, in that
+// case when the index represented by selBits is greater than or equal with
+// len(inputs), the output will be undefined.
+//
+// BinaryMux does NOT define any boolean constraints for selBits. Outside this
+// function, all the elements of selBits must be constrained to be binary
+// digits.
+func BinaryMux(api frontend.API, selBits, inputs []frontend.Variable) frontend.Variable {
+	// The number of defined R1CS constraints for an input of length n is always n - 1.
+	// n does not need to be a power of 2.
+	if len(selBits) == 0 {
+		return inputs[0]
+	}
+
+	nextSelBits := selBits[:len(selBits)-1]
+	sel := selBits[len(selBits)-1]
+	pivot := 1 << len(nextSelBits)
+	if pivot >= len(inputs) {
+		return BinaryMux(api, nextSelBits, inputs)
```

Comment:
If we change this line to this
```go
return api.Mul(api.Sub(1, sel), BinaryMux(api, nextSelBits, inputs))
```
We will not have undefined behavior anymore, and the return value will be zero when the selector is outside. However it is not as efficient as using a comparison.

Reply:
I think the previous approach is better (asserting input len is power of two)